File: ta_lib/strategies/pulse/src/adx.rs
use base::prelude::*;
use core::prelude::*;
use momentum::dmi;
use timeseries::prelude::*;

const ADX_THRESHOLD: f32 = 30.;

pub struct AdxPulse {
    smooth: Smooth,
    period_adx: usize,
    period_di: usize,
    threshold: f32,
}

impl AdxPulse {
    pub fn new(smooth: Smooth, period_adx: f32, period_di: f32, threshold: f32) -> Self {
        Self {
            smooth,
            period_adx: period_adx as usize,
            period_di: period_di as usize,
            threshold,
        }
    }
}

impl Pulse for AdxPulse {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period_adx, self.period_di)
    }

    fn assess(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let atr = data.atr(self.smooth, self.period_di);
        let (_, _, adx) = dmi(
            data.high(),
            data.low(),
            &atr,
            self.smooth,
            self.period_adx,
            self.period_di,
        );

        let barrier = ADX_THRESHOLD - self.threshold;

        (adx.sgt(&barrier), adx.sgt(&barrier))
    }
}

File: strategy/generator/pulse/adx.py
from dataclasses import dataclass

from core.models.parameter import Parameter, RandomParameter, StaticParameter
from core.models.smooth import Smooth

from .base import Pulse, PulseType


@dataclass(frozen=True)
class AdxPulse(Pulse):
    type: PulseType = PulseType.Adx
    smooth: Parameter = StaticParameter(Smooth.SMMA)
    period: Parameter = StaticParameter(15.0)
    period_di: Parameter = StaticParameter(15.0)
    threshold: Parameter = RandomParameter(0.0, 5.0, 1.0)
